File: electron-gui/server/src/soffice.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { execSync } from 'node:child_process'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

const WINDOWS_CANDIDATES = [
  'C:\\Program Files\\LibreOffice\\program\\soffice.exe',
  'C:\\Program Files (x86)\\LibreOffice\\program\\soffice.exe',
]

const MACOS_CANDIDATES = ['/Applications/LibreOffice.app/Contents/MacOS/soffice']

const LINUX_CANDIDATES = ['/usr/bin/soffice', '/usr/lib/libreoffice/program/soffice', '/opt/libreoffice/program/soffice']

// 內建可攜版相對於 server bundle 的位置：
// - 開發模式：server/src/soffice.ts -> ../../resources/libreoffice-portable
// - 打包後：  resources/server/dist/main.js -> ../../libreoffice-portable
const PORTABLE_CANDIDATES = [
  path.join(__dirname, '..', '..', 'resources', 'libreoffice-portable', 'program', 'soffice.exe'),
  path.join(__dirname, '..', '..', 'libreoffice-portable', 'program', 'soffice.exe'),
]

export interface SofficeLocation {
  path: string
  portable: boolean
}

let cached: SofficeLocation | null | undefined

export function findSofficePath(): SofficeLocation | null {
  if (cached !== undefined) return cached

  try {
    const which = os.platform() === 'win32' ? 'where soffice' : 'which soffice'
    const found = execSync(which, { encoding: 'utf-8' }).trim().split('\n')[0]?.trim()
    if (found && fs.existsSync(found)) {
      cached = { path: found, portable: false }
      return cached
    }
  } catch {
    // not on PATH
  }

  const candidates =
    os.platform() === 'win32' ? WINDOWS_CANDIDATES : os.platform() === 'darwin' ? MACOS_CANDIDATES : LINUX_CANDIDATES

  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) {
      cached = { path: candidate, portable: false }
      return cached
    }
  }

  if (os.platform() === 'win32') {
    for (const candidate of PORTABLE_CANDIDATES) {
      if (fs.existsSync(candidate)) {
        cached = { path: candidate, portable: true }
        return cached
      }
    }
  }

  cached = null
  return cached
}

export const SOFFICE_NOT_FOUND_MESSAGE =
  '找不到 LibreOffice（soffice）。此功能需要安裝 LibreOffice 才能將 PPT/PPTX 轉換為圖片，請至 https://www.libreoffice.org/download/ 安裝後重試。'

File: soffice.py
from __future__ import annotations

import platform
import shutil
from pathlib import Path

# 對齊 GUI 版 soffice.ts 的候選路徑（不含 electron 打包的 portable 版）
_WINDOWS_CANDIDATES = [
    r"C:\Program Files\LibreOffice\program\soffice.exe",
    r"C:\Program Files (x86)\LibreOffice\program\soffice.exe",
]

_MACOS_CANDIDATES = ["/Applications/LibreOffice.app/Contents/MacOS/soffice"]

_LINUX_CANDIDATES = [
    "/usr/bin/soffice",
    "/usr/lib/libreoffice/program/soffice",
    "/opt/libreoffice/program/soffice",
]

SOFFICE_NOT_FOUND_MESSAGE = (
    "找不到 LibreOffice（soffice）。此功能需要安裝 LibreOffice 才能將 PPT/PPTX 轉換為圖片，"
    "請至 https://www.libreoffice.org/download/ 安裝後重試。"
)

_UNSET = object()  # sentinel：尚未偵測
_cached: str | None | object = _UNSET


def find_soffice() -> str | None:
    """偵測 soffice 執行檔路徑，找不到回傳 None。結果會快取。"""
    global _cached
    if _cached is not _UNSET:
        return _cached  # type: ignore[return-value]

    system = platform.system()
    if system == "Windows":
        candidates = _WINDOWS_CANDIDATES
    elif system == "Darwin":
        candidates = _MACOS_CANDIDATES
    else:
        candidates = _LINUX_CANDIDATES

    # Windows 上 shutil.which("soffice") 會依 PATHEXT 命中 soffice.COM（GUI 包裝器），
    # 但 --convert-to 應使用 soffice.exe，故優先找 .exe。
    if system == "Windows":
        found = shutil.which("soffice.exe")
        if found and Path(found).is_file():
            _cached = found
            return _cached
    else:
        found = shutil.which("soffice")
        if found and Path(found).is_file():
            _cached = found
            return _cached

    for candidate in candidates:
        if Path(candidate).is_file():
            _cached = candidate
            return _cached

    _cached = None
    return _cached
